pkg/pgmodel/ingestor/sql_ingest.go

```diff
@@ -0,0 +1,276 @@
+
+	conn := &utils.PgxConnImpl{
+		Conn: c,
+	}
+
+	pi, err := newPgxInserter(conn, cache, Config)
+	if err != nil {
+		return nil, err
+	}
+
+	return &DBIngestor{db: pi}, nil
+}
+
+func newPgxInserter(conn utils.PgxConn, cache utils.MetricCache, Config *Config) (*pgxInserter, error) {
+	cmc := make(chan struct{}, 1)
+
+	numCopiers := Config.NumCopiers
+	if numCopiers < 1 {
+		log.Warn("msg", "num copiers less than 1, setting to 1")
+		numCopiers = 1
+	}
+
+	// We run inserters bus-style: all of them competing to grab requests off a
+	// single channel. This should offer a decent compromise between batching
+	// and balancing: if an inserter is awake and has little work, it'll be more
+	// likely to win the race, while one that's busy or asleep won't.
+	toCopiers := make(chan copyRequest, numCopiers)
+	for i := 0; i < numCopiers; i++ {
+		go runInserter(conn, toCopiers)
+	}
+
+	inserter := &pgxInserter{
+		conn:                   conn,
+		metricTableNames:       cache,
+		completeMetricCreation: cmc,
+		asyncAcks:              Config.AsyncAcks,
+		toCopiers:              toCopiers,
+	}
+	if Config.AsyncAcks && Config.ReportInterval > 0 {
+		inserter.insertedDatapoints = new(int64)
+		reportInterval := int64(Config.ReportInterval)
+		go func() {
+			log.Info("msg", fmt.Sprintf("outputting throughpput info once every %ds", reportInterval))
+			tick := time.Tick(time.Duration(reportInterval) * time.Second)
+			for range tick {
+				inserted := atomic.SwapInt64(inserter.insertedDatapoints, 0)
+				log.Info("msg", "Samples write throughput", "samples/sec", inserted/reportInterval)
+			}
+		}()
+	}
+	//on startup run a completeMetricCreation to recover any potentially
+	//incomplete metric
+	err := inserter.CompleteMetricCreation()
+	if err != nil {
+		return nil, err
+	}
+
+	go inserter.runCompleteMetricCreationWorker()
+
+	return inserter, nil
+}
+
+type pgxInserter struct {
+	conn                   utils.PgxConn
+	metricTableNames       utils.MetricCache
+	inserters              sync.Map
+	completeMetricCreation chan struct{}
+	asyncAcks              bool
+	insertedDatapoints     *int64
+	toCopiers              chan copyRequest
+}
+
+func (p *pgxInserter) CompleteMetricCreation() error {
+	_, err := p.conn.Exec(
+		context.Background(),
+		finalizeMetricCreation,
+	)
+	return err
+}
+
+func (p *pgxInserter) runCompleteMetricCreationWorker() {
+	for range p.completeMetricCreation {
+		err := p.CompleteMetricCreation()
+		if err != nil {
+			log.Warn("msg", "Got an error finalizing metric", "err", err)
+		}
+	}
+}
+
+func (p *pgxInserter) Close() {
+	close(p.completeMetricCreation)
+	p.inserters.Range(func(key, value interface{}) bool {
+		close(value.(chan insertDataRequest))
+		return true
+	})
+	close(p.toCopiers)
+}
+
+func (p *pgxInserter) InsertNewData(rows map[string][]utils.SamplesInfo) (uint64, error) {
+	return p.InsertData(rows)
+}
```

Comment:
It looks like this func is not needed

Reply:
Yeah, but I think that's basically for running unit-tests.